content/blog/2019-09-25-RecordTypes.md

```diff
@@ -0,0 +1,131 @@
+`<field# value>` is an identifier for an existing variable matching the field type
+
+We decided to introduce the `record` keyword to match the precedent of using `id` or `const` in front of an identifier or constant value. Note that the ordering of field name and value pairs do not matter, as long as they match with the definiton’s field names and types. We also only allow field values to be existing variables to match the semantics of current operations. The structure of this statement was designed to match record type declarations as much as possible. 
+
+
+### Access
+We use the dot operator to access a field of a record:
+
+    <new record name> : <type> = <record> . <field>
+
+Where:
+`<new record name>` is a valid identifier that takes on the value of the indicated field
+`<record>` is the name of an instance of a record with a field name `<field>` that has type `<type>`
+Note that there is a space before and after the dot. This is strictly necessary as dot is a valid character for a variable name and we decided that it would be horrible for backwards compatibility if we changed variable naming rules.
+
+
+We chose this format because using the dot operator to access fields is very common in modern programming languages.
+
+### *With* Syntax
+While immutable data structures allow you to more easily reason about how values flow through your program, these currently immutable records are cumbersome to change. 
+For example, to change one field, you must recreate the entire record. While we do want a new record to be created, copying variables between two records is overly tedious. 
+As such, we decided to implement a with syntax, similar to OCaml in which the user just needs to specify a record name as the base, and then the fields desired to be changed, along with the new value. This maintains immutability, without so much tedium.
+
+We use the following syntax:
+
+    <new record>: <record type> = 
+        <old record> with {<field1 name>: <field1 value>; <field2 name>: <field2 value> … };
+
+Where:
+`<new record>` is the name of the new record
+`<record type>` is the same type as `<old record>`
+`with` is a new keyword
+`<field# name>` must match with a field in `<type>`
+`<field# value>` must be a variable with a type that matches its field name.
+Within the braces, the user may specify 0 to n field name and value pairs, where n is the total number of fields in the record type.
+
+This syntax was designed to have a similar format as record instantiation. 
+
+### Immutability
+Immutable data structures are easier to reason about when developing compiler optimizations. However, forcing developers to recreate the entire record every time they need to change a value poses usability challenges. We decided on sticking to immutability in part for optimizations' sake. The ability to make guarantees that a value will not change lends itself well to constant-folding, even after a function call that takes in your record as an argument. 
```

Comment:
Because this is a top-level design concern, it might be wise to put it at the top of your writeup. Also, I’d argue that usability is not a paramount concern for a compiler IR—but the ability to compile higher-level languages into it is.